core/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -311,6 +313,14 @@ public Node start() {
         // Start the transport service now so the publish address will be added to the local disco node in ClusterService
         TransportService transportService = injector.getInstance(TransportService.class);
         transportService.start();
+        DiscoveryNode localNode = injector.getInstance(DiscoveryNodeService.class)
+                .buildLocalNode(transportService.boundAddress().publishAddress());
+        clusterService.setLocalNode(localNode);
+        transportService.setLocalNode(localNode);
+        clusterService.add(transportService.getTaskManager());
+
+        // state NodeConnectionService last, so it's listeners will be first and last in cluster service
```

Comment:
for instance in RestoreService we use `addLast` at runtime which messes with this assumption, that entire order thing is broken and error prone. The best thing I can come up with so far is to add defined stage like this:

```Java

enum ApplyStage {
  NewClusterState,
  NodesConnected,
  StateRecovered,
  ShardsStarted,
  RepositoriesCreated,
  NodesDisconnected;
}
```
where listeners can be registered but I am not too happy about it...